README.md
# cti-stix-elevator

*This is an [OASIS Open Repository](https://www.oasis-open.org/resources/open-repositories/). See the [Governance](#governance) section for more information.*

The stix-elevator is a software tool for converting STIX 1.2 XML to STIX 2.0 JSON. Due to the differences between STIX 1.2 and STIX 2.0, this conversion is best-effort only, and stix-elevator cannot convert from STIX 2.0 back to STIX 1.2. During the conversion, stix-elevator provides information on the assumptions it needs to make to produce valid STIX 2.0 JSON, and what information was not able to be converted.

The stix-elevator is a work-in-progress.  It should be used to explore how existing STIX 1.x would potentially be represented in STIX 2.0.  Using the current version of the elevator will provide insight to issues that might need to be mitigated to convert your STIX 1.x content.

**_It should not be used in a production environment, and should not be considered final._**

**_STIX 1.x Composite Indicator Expressions and CybOX 2.x Composite Observable Expressions allow a level of flexibility not present in STIX 2 patterns. These composite expressions can frequently have ambiguous interpretations, so STIX 2 Indicators created by the stix-elevator from STIX 1.x Indicators containing composite expressions should be inspected to ensure the STIX 2 Indicator has the intended meaning._**

Please enter any comments on how to improve it into the issue tracker.

## Requirements

For STIX 1.1.1 content;

* Python 2.6/2.7
* python-stix >= 1.1.1.7 (other dependencies inherited from python-stix)

For STIX 1.2 content:

* Python 3.5
* python-stix >= 1.2.0.0 (other dependencies inherited from python-stix)

For both:

* stix2validator >= 0.2.0 (including its dependencies:  jsonschema, colorama, requests)
* pycountry >= 1.20


## Installation

stix-elevator should be installed using the git repository:

```
$ pip install git+https://github.com/oasis-open/cti-stix-elevator.git
```

This will install all necessary dependencies.

### Using pypi:

For STIX 1.1.1 content:

```
$ pip install stix<1.2  - this will install python-stix 1.1.1.8
$ pip install stix2-elevator
```

For STIX 1.2 content:

```
$ pip install stix2-elevator  - this will install python-stix 1.2.0.2

```

## Usage


### As A Script

The elevator comes with a bundled script which you can use to elevate STIX 1.1.1 - 1.2.1 content to STIX 2.0 content:

```
usage: stix2_elevator [-h] [--incidents] [--no-squirrel-gaps]
                      [--infrastructure]
                      [--package-created-by-id PACKAGE_CREATED_BY_ID]
                      [--default-timestamp DEFAULT_TIMESTAMP]
                      [--validator-args VALIDATOR_ARGS] [-e ENABLE]
                      [-d DISABLE] [-s]
                      [--message-log-directory MESSAGE_LOG_DIRECTORY]
                      file

positional arguments:
  file                  The input STIX 1.x document to be elevated.

optional arguments:
  -h, --help            show this help message and exit

  --incidents           Incidents will be included in the conversion.

  --no-squirrel-gaps    Do not include STIX 1.x content that cannot be
                        represented directly in STIX 2.0 using the description
                        property.

  --infrastructure      Infrastructure will be included in the conversion.

  --package-created-by-id PACKAGE_CREATED_BY_ID
                        Use provided identifier for "created_by_ref"
                        properties.Example: --package-created-by-id "identity
                        --1234abcd-1a12-12a3-0ab4-1234abcd5678"

  --default-timestamp DEFAULT_TIMESTAMP
                        Use provided timestamp for properties that require a
                        timestamp. Example: --default-timestamp
                        "2016-11-15T13:10:35.053000Z"

  --validator-args VALIDATOR_ARGS
                        Arguments to pass stix-validator. Default: --strict-
                        types Example: stix_elevator.py <file> --validator-
                        args "-v --strict-types -d 212"

  -e ENABLE, --enable ENABLE
                        A comma-separated list of the elevator messages to
                        enable. If the --disable option is not used, no other
                        messages will be shown. Example: stix_elevator.py
                        <file> --enable 250

  -d DISABLE, --disable DISABLE
                        A comma-separated list of the elevator messages to
                        disable. Example: stix_elevator.py <file> --disable
                        212,220

  -s, --silent          If this flag is set. All elevator messages will be
                        disabled.

  --message-log-directory MESSAGE_LOG_DIRECTORY
                        If this flag is set. All elevator messages will be
                        saved to file. The name of the file will be the input
                        file with extension .log in the specified directory.
                        Note, make surethe directory already exists. Example:
                        stix_elevator.py <file> --message-log-directory
                        "..\logs"
```

Refer to elevator_log_messages.xlsx for error codes.

### As A Library

You can also use this library to integrate STIX elevation into your own tools. You can elevate a STIX 1.x file:

```
  from stix2elevator import elevate_file

  results = elevate_file("stix_file.xml")
  print(results)
```
Additionally, a similar method exists to accept a string as an argument:

```
  from stix2elevator import elevate_string

  results = elevate_string("...")
  print(results)
```

## Governance

This GitHub public repository ( **[https://github.com/oasis-open/cti-stix-elevator](https://github.com/oasis-open/cti-stix-elevator)** ) was [proposed](https://lists.oasis-open.org/archives/cti/201610/msg00106.html) and [approved](https://lists.oasis-open.org/archives/cti/201610/msg00126.html) [[bis](https://issues.oasis-open.org/browse/TCADMIN-2477)] by the [OASIS Cyber Threat Intelligence (CTI) TC](https://www.oasis-open.org/committees/cti/) as an [OASIS Open Repository](https://www.oasis-open.org/resources/open-repositories/) to support development of open source resources related to Technical Committee work.

While this Open Repository remains associated with the sponsor TC, its development priorities, leadership, intellectual property terms, participation rules, and other matters of governance are [separate and distinct](https://github.com/oasis-open/cti-stix-elevator/blob/master/CONTRIBUTING.md#governance-distinct-from-oasis-tc-process) from the OASIS TC Process and related policies.

All contributions made to this Open Repository are subject to open source license terms expressed in the [BSD-3-Clause License](https://www.oasis-open.org/sites/www.oasis-open.org/files/BSD-3-Clause.txt). That license was selected as the declared ["Applicable License"](https://www.oasis-open.org/resources/open-repositories/licenses) when the Open Repository was created.

As documented in ["Public Participation Invited](https://github.com/oasis-open/cti-stix-elevator/blob/master/CONTRIBUTING.md#public-participation-invited)", contributions to this OASIS Open Repository are invited from all parties, whether affiliated with OASIS or not. Participants must have a GitHub account, but no fees or OASIS membership obligations are required. Participation is expected to be consistent with the [OASIS Open Repository Guidelines and Procedures](https://www.oasis-open.org/policies-guidelines/open-repositories), the open source [LICENSE](https://github.com/oasis-open/cti-stix-elevator/blob/master/LICENSE) designated for this particular repository, and the requirement for an [Individual Contributor License Agreement](https://www.oasis-open.org/resources/open-repositories/cla/individual-cla) that governs intellectual property.

### <a id="maintainers">Maintainers</a>

Open Repository [Maintainers](https://www.oasis-open.org/resources/open-repositories/maintainers-guide) are responsible for oversight of this project's community development activities, including evaluation of GitHub [pull requests](https://github.com/oasis-open/cti-stix-elevator/blob/master/CONTRIBUTING.md#fork-and-pull-collaboration-model) and [preserving](https://www.oasis-open.org/policies-guidelines/open-repositories#repositoryManagement) open source principles of openness and fairness. Maintainers are recognized and trusted experts who serve to implement community goals and consensus design preferences.

Initially, the associated TC members have designated one or more persons to serve as Maintainer(s); subsequently, participating community members may select additional or substitute Maintainers, per [consensus agreements](https://www.oasis-open.org/resources/open-repositories/maintainers-guide#additionalMaintainers).

**<a id="currentMaintainers">Current Maintainers of this Open Repository</a>**

*   [Greg Back](mailto:gback@mitre.org); GitHub ID: [https://github.com/gtback/](https://github.com/gtback/); WWW: [MITRE](https://www.mitre.org/)
*   [Rich Piazza](mailto:rpiazza@mitre.org); GitHub ID: [https://github.com/rpiazza/](https://github.com/rpiazza/); WWW: [MITRE](https://www.mitre.org/)

## <a id="aboutOpenRepos">About OASIS Open Repositories</a>

*   [Open Repositories: Overview and Resources](https://www.oasis-open.org/resources/open-repositories/)
*   [Frequently Asked Questions](https://www.oasis-open.org/resources/open-repositories/faq)
*   [Open Source Licenses](https://www.oasis-open.org/resources/open-repositories/licenses)
*   [Contributor License Agreements (CLAs)](https://www.oasis-open.org/resources/open-repositories/cla)
*   [Maintainers' Guidelines and Agreement](https://www.oasis-open.org/resources/open-repositories/maintainers-guide)

## <a id="feedback">Feedback</a>

Questions or comments about this Open Repository's activities should be composed as GitHub issues or comments. If use of an issue/comment is not possible or appropriate, questions may be directed by email to the Maintainer(s) [listed above](#currentMaintainers). Please send general questions about Open Repository participation to OASIS Staff at [repository-admin@oasis-open.org](mailto:repository-admin@oasis-open.org) and any specific CLA-related questions to [repository-cla@oasis-open.org](mailto:repository-cla@oasis-open.org).


CHANGELOG
CHANGELOG
=========

4.1.7 - 2022-09-28

    - A report or grouping is created to store the extra STIX package header information

4.1.6 - 2022-06-10

    - fix functionality for --enable and --disable options

4.1.5 - 2022-02-04

    - Additional fix for Hostname in SocketAddress in pattern

4.1.4 - 2022-01-24

    - Handle required properties
        - new command line option -r
    - Hostname in SocketAddress in pattern is fixed
    - Revamp handling of Sightings
    - Compare STIX 1.x ids in a case-insensitive way

4.1.3 - 2022-01-11

    ACS Data Marking issues
    - Incorporate change to package dependency stix-edh
    - Warn if ACS identifier is invalid

4.1.2 - 2021-12-22

    - Improved processing of ACS data markings
    - Handle multiple Vulnerabilites in an Exploit-Target correctly

4.1.1 - 2021-12-09

Minor fixes to handle an empty STIX 1.x description tag, and improved processing of extensions in patterns.

4.1.0 - 2021-10-08

This release optionally handles ACS data markings (see install doc)

* Other changes

    - added --acs oiption for ACS data markings
    - Fix hash_constant processing to handle defaults
    - Handle literals in missing policy code
    - Added time properties to incident extension
    - Handle malware aliases

4.0.2 - 2021-09-10

* Changes

    - Avoid generating nested observation expressions, which are invalid
    - Handle cybox Homename object as domain-name object in STIX 2.x
    - Make sure identity created_by_ref property is set correctly
    - Correcly convert file object properties in patterns
    - Handle nested related objects in STIX 1.x Observables
    - Handle Information Source in a STIX 1.x package data marking
    - Better way to avoid infinite recursion when called get_identity
    - Add suffix 'ref(s)'' to property names for custom/extension of Incident
    - AIS data markings:
        - Insure TTP generated STIX 2.x objects and relationship objects have the correct AIS data markings
        - Change labels to handle CISA-PROPRIETARY correctly

4.0.1 - 2021-06-10

Default spec_version when invoking the elevator programmatically (i.e., not from the command line)
has been corrected to be 2.1

4.0.0 - 2021-05-02

The elevator and stepper are compliant with STIX 2.1 CS03

Python 3.5 is no longer supported.
All compatibility code to support Python 2.7 has been removed (e.g., the six package)

* Major Changes
    - Support Extension feature described in section 7.3 of the specification
    - Support Incident as a specification SDO
    - Use handle_missing_properties method to encapsulate most extension/custom functionality
    - Handle related objects with no know STIX 2.1 relationship

* Other Changes
    - Handle ArchiveFile different for 2.0 vs 2.1
    - email_message:message_id is only in 2.1
    - Infer network-traffic protocol from extension type
    - Handle registry key in example where there is no hive property
    - Better handling of Sightings
    - Add object data marking to Relationships when appropriate based on the source and target references
    - Handle name and title properties of malware better

* Testing changes
    - Added testing for python 3.9
    - Test for the ignore missing policy for STIX 2.1

3.0.1 - 2020-12-09

   * Update dependencies
        - python-stix2 (2.1.0)
        - stix2validator (2.0.2)
        - stixmarx (1.0.8)

   * Other changes:
        - Implement semantics of patterns more correctly
        - Impose predicable order for terms in pattern expression involving pdf files
        - Clear location object cache


3.0.0 - 2020-07-08

    **The elevator and stepper are compliant with STIX 2.1 CS01**

    **Python 2.7 and 3.4 are no longer supported**

    * Other changes:
        - Handle STIX 1.x (CybOX 2.1) custom cyber objects
        - Handle SCOs: x509-certificate, autonomous-system, software
        - Handle File extensions: raster-image-ext, pdf-ext
        - Correct deterministic id generation algorithm
        - Handle observable characteristics in infrastructure
        - Handle parameter observables in COA
        - *_types property is now optional in 2.1
        - Handle some STIX 1.x object references
        - Improved handling of STIX 1.x RelatedObjects
        - Handle aliases for threat actors better
        - Threat_actor goal set correctly
        - Handle socket address
        - Handle more headers in email-message
        - Improved handling of composite observable conversion
        - Add general elevate() method, deprecate all other methods
        - Consolidate timestamp logic
        - Use uuid4s for custom SCOs in the stepper
        - Check stepper results with stix-validator
        - In external-references, the reference property must be in URI format

    * Testing changes
        - Added testing for python 3.8
        - Add new file to test main methods to interact with the elevator, use new method in test_idioms.py

    * Fixes for issues:
        #174 - In test case identifying-a-threat-actor-group, a uuid is reused inappropriately
        #182 - Message 905 about package timestamp is misleading
        #186 - generate_sco_id fails to generate deterministic ID for some objects that contain special characters
        #191 - Required attributes not added when upgrading from v2.0 -> v2.1
        #193 - 'MarkableText' object has no attribute 'reference'

2.1.1 - 2019-12-20

    * Handle archive files correctly
    * Handle CIDRs
    * Handle missing kill-chain definition

2.1.0 - 2019-12-11

    *	Handle SCOs
    *	More complete support of CybOX objects
    *	Enable use of custom properties for properties missing from STIX 2.x
    *	Support all additional properties and property name changes for STIX 2.1 WD06
    *	Handle UUIDv5 for SCOs in STIX 2.1

    *	Fix patterns involving PE binary file header
    *	Handle characteristic observables in infrastructure
    *	Better mapping of STIX 1.x relationship types to STIX 2.x ones

stix2elevator/cli.py
#!/usr/bin/env python

"""The stix2-elevator is a work-in-progress. It should be used to explore how
existing STIX 1.x would potentially be represented in STIX 2.x. Using the
current version of the stix2-elevator will provide insight to issues that might
need to be mitigated to convert your STIX 1.x content.
"""

# Standard Library
import argparse
import sys

# internal
from stix2elevator import elevate
from stix2elevator.options import initialize_options
from stix2elevator.utils import NewlinesHelpFormatter
from stix2elevator.version import __version__

CODE_TABLE = """
Refer to elevator_log_messages.rst for all stix2-elevator messages. Use the
associated code number to --enable or --disable a message. By default, the
stix2-elevator displays all messages. Note: disabling the message does not
disable the functionality.
"""


def _get_arg_parser(is_script=True):
    """Create and return an ArgumentParser for this application."""

    desc = "stix2-elevator v{0}\n\n".format(__version__)

    parser = argparse.ArgumentParser(
        description=desc + __doc__,
        formatter_class=NewlinesHelpFormatter,
        epilog=CODE_TABLE
    )

    if is_script:
        parser.add_argument(
            "file_",
            help="The input STIX 1.x document to be elevated.",
            metavar="file"
        )

    parser.add_argument(
        "--missing-policy",
        help="Policy for including STIX 1.x content that cannot be represented "
             "directly in STIX 2.x.  The default is 'add-to-description'.",
        choices=["use-custom-properties", "add-to-description", "ignore", "use-extensions"],
        dest="missing_policy",
        action="store",
        default="add-to-description"
    )

    parser.add_argument(
        "--header-object-type",
        help="What STIX 2 type to use to store extra information from the STIX 1 package header."
             "The default is 'report'.",
        choices=["report", "grouping"],
        dest="header_object_type",
        action="store",
        default="report"
    )

    parser.add_argument(
        "--custom-property-prefix",
        help="Prefix to use for custom property names when missing policy is 'use-custom-properties'. The default is 'elevator'.",
        dest="custom_property_prefix",
        action="store",
        default="elevator"
    )

    parser.add_argument(
        "--infrastructure",
        help="Infrastructure will be included in the conversion. Default for version 2.1 is true.",
        dest="infrastructure",
        action="store_true",
        default=False
    )

    parser.add_argument(
        "--incidents",
        help="Incidents will be included in the conversion.  Default for version 2.1 is true.",
        dest="incidents",
        action="store_true",
        default=False
    )

    parser.add_argument(
        "--acs",
        help="Process ACS data markings",
        dest="acs",
        action="store_true",
        default=False
    )

    parser.add_argument(
        "--package-created-by-id",
        help="Use provided identifier for \"created_by_ref\" properties. "
             "Example: --package-created-by-id \"identity--1234abcd-1a12-42a3-0ab4-1234abcd5678\"",
        dest="package_created_by_id",
        action="store",
        default=None
    )

    parser.add_argument(
        "--default-timestamp",
        help="Use provided timestamp for properties that require a timestamp. "
             "\n\nExample: --default-timestamp \"2016-11-15T13:10:35.053000Z\"",
        dest="default_timestamp",
        action="store",
        default=None
    )

    parser.add_argument(
        "--validator-args",
        help="Arguments to pass to stix2-validator. Default: --strict-types\n\n"
             "Example: stix2_elevator.py <file> --validator-args=\"-v --strict-types -d 212\"",
        dest="validator_args",
        action="store",
        default=""
    )

    parser.add_argument(
        "-e",
        "--enable",
        help="A comma-separated list of the stix2-elevator messages to enable. "
             "Not to be used with --disable. \n\nExample: stix2_elevator.py <file> --enable 250",
        dest="enabled",
        default=None
    )

    parser.add_argument(
        "-d",
        "--disable",
        help="A comma-separated list of the stix2-elevator messages to disable. \n\n"
             "Not to be used with --enable. \n\nExample: stix2_elevator.py <file> --disable 212,220",
        dest="disabled",
        default=None
    )

    parser.add_argument(
        "-s",
        "--silent",
        help="If this flag is set, all stix2-elevator messages will be disabled.",
        dest="silent",
        action="store_true",
        default=False
    )

    parser.add_argument(
        "--message-log-directory",
        help="If this flag is set, all stix2-elevator messages will be saved to "
             "file. The name of the file will be the input file with "
             "extension .log in the specified directory. Note, make sure "
             "the directory already exists.\n\n"
             "Example: stix2_elevator.py <file> --message-log-directory \"../logs\"",
        dest="message_log_directory",
        action="store",
        default=None
    )

    parser.add_argument(
        "--log-level",
        default="INFO",
        help="The logging output level.",
        choices=["DEBUG", "INFO", "WARN", "ERROR", "CRITICAL"]
    )

    parser.add_argument(
        "-m",
        "--markings-allowed",
        help="Avoid an error exit, if these markings are in the content, but not supported by the elevator.  \n\n"
             "Specify as a comma-separated list"
             "Example: stix2_elevator.py < file > --markings-allowed \"ISAMarkingsAssertion,ISAMarkings\"",

        dest="markings_allowed",
        action="store",
        default=""
    )

    parser.add_argument(
        "-p",
        "--error-policy",
        "--policy",     # deprecated
        help="The policy to deal with errors.  The default is 'no_policy'.",
        dest="policy",
        choices=["no_policy", "strict_policy"],
        action="store",
        default="no_policy"
    )

    parser.add_argument(
        "-v",
        "--version",
        help="The version of stix 2 to be produced.  The default is 2.1",
        dest="spec_version",
        choices=["2.0", "2.1"],
        action="store",
        default="2.1"
    )

    parser.add_argument(
        "-r",
        "--ignore-required-properties",
        help="Do not provide missing required properties",
        dest="ignore_required_properties",
        action="store_true",
        default=False
    )

    return parser


def main():
    # Parse stix2-elevator command-line args
    elevator_parser = _get_arg_parser()
    elevator_args = elevator_parser.parse_args()
    sys.setrecursionlimit(3000)
    if initialize_options(options=elevator_args):
        result = elevate(elevator_args.file_)
        if result:
            sys.stdout.write(result + "\n")
        else:
            sys.exit(1)
    else:
        sys.exit(1)


if __name__ == '__main__':
    main()
